src/main/groovy/com/morpheus/scvmm/sync/VirtualMachineSync.groovy

```diff
@@ -0,0 +1,490 @@
+        try {
+            for (cloudItem in addList) {
+                log.debug "Adding new virtual machine: ${cloudItem.Name}"
+                def vmConfig = buildVmConfig(cloudItem, defaultServerType)
+                ComputeServer add = new ComputeServer(vmConfig)
+                add.maxStorage = (cloudItem.TotalSize?.toDouble() ?: 0)
+                add.usedStorage = (cloudItem.UsedSize?.toDouble() ?: 0)
+                add.maxMemory = (cloudItem.Memory?.toLong() ?: 0) * 1024l * 1024l
+                add.maxCores = cloudItem.CPUCount.toLong() ?: 1
+                add.parentServer = hosts?.find { host -> host.externalId == cloudItem.HostId }
+                add.plan = SyncUtils.findServicePlanBySizing(availablePlans, add.maxMemory, add.maxCores, null, fallbackPlan, null, cloud.account, availablePlanPermissions)
+                if (cloudItem.IpAddress) {
+                    add.externalIp = cloudItem.IpAddress
+                }
+                if (cloudItem.InternalIp) {
+                    add.internalIp = cloudItem.InternalIp
+                }
+                // Operating System
+                def osTypeCode = apiService.getMapScvmmOsType(cloudItem.OperatingSystem, true, cloudItem.OperatingSystemWindows?.toString() == 'true' ? 'windows' : null)
+                def osTypeCodeStr = osTypeCode ?: 'other'
+                def osType = context.services.osType.find(new DataQuery().withFilter('code', osTypeCodeStr))
+                if (osType) {
+                    add.serverOs = osType
+                    add.osType = add.serverOs?.platform?.toLowerCase()
+                    add.platform = osType?.platform
+                }
+                add.sshHost = add.internalIp ?: add.externalIp
+                if (consoleEnabled) {
+                    add.consoleType = 'vmrdp'
+                    add.consoleHost = add.parentServer?.name
+                    add.consolePort = 2179
+                    add.sshUsername = cloud.accountCredentialData?.username ?: cloud.getConfigProperty('username') ?: 'dunno'
+                    if (add.sshUsername.contains('\\')) {
+                        add.sshUsername = add.sshUsername.tokenize('\\')[1]
+                    }
+                    add.consolePassword = cloud.accountCredentialData?.password ?: cloud.getConfigProperty('password')
+                }
+                add.capacityInfo = new ComputeCapacityInfo(maxCores: add.maxCores, maxMemory: add.maxMemory, maxStorage: add.maxStorage)
+                ComputeServer savedServer = context.async.computeServer.create(add).blockingGet()
+                if (!savedServer) {
+                    log.error "error adding new virtual machine: ${add}"
+                } else {
+                    syncVolumes(savedServer, cloudItem.Disks)
```

Comment:
sync volumes adds new volumes to the server, saves the server, and then the server is resaved after the volume sync? This is calling the API more than it needs to and creates the potential for overwriting data between all the saves. There are also no bulk actions being leveraged here.